chebi-update/src/main/java/org/reactome/release/chebiupdate/ChebiDataRetriever.java

```diff
@@ -0,0 +1,267 @@
+	{
+		String identifier = (String) molecule.getAttributeValue(ReactomeJavaConstants.identifier);
+		// If there's a valid cache and we're supposed to use it, log a message here indicating
+		// that there must have been a cache miss.
+		if (this.useCache && chebiCache.size() > 0)
+		{
+			logger.trace("Cache miss for CHEBI:{}", identifier);
+		}
+		Entity entity = this.chebiClient.getCompleteEntity(identifier);
+		// IF there is valid entity AND we are supposed to use a cache, then write the entity to the cache file.
+		if (entity != null && this.useCache)
+		{
+			bw.write("CHEBI:"+identifier+"\t"+entity.getChebiId()+"\t"+entity.getChebiAsciiName()+"\t"+ (entity.getFormulae().size() > 0 ? entity.getFormulae().get(0).getData() : "") + "\t" + LocalDateTime.now().toString() + "\n");
+			bw.flush();
+		}
+		// IF the entity was null, then log a message that the webservice response was null!
+		if (entity == null)
+		{
+			failedEntitiesList.put(molecule, "ChEBI WebService response was NULL.");
+		}
+		
+		return entity;
+	}
+
+	/**
+	 * Returns the data from the cache file (if it exists). If it doesn't exit, then an empty map will be returned.
+	 * @return A mapping of ChEBI ID to a list with items in this order: ChEBI (might be different if ChEBI has changed identifiers), ChEBI Name, Formula.
+	 * @throws IOException
+	 */
+	private Map<String,List<String>> loadCacheFromFile() throws IOException
+	{
+		Map<String,List<String>> chebiCache = Collections.synchronizedMap(new HashMap<String, List<String>>());
+		if (this.useCache)
+		{
+			logger.info("useCache is TRUE - chebi-cache file will be read, and populated. Identifiers not in the cache will be queried from ChEBI.");
+			// if the cache exists, load it.
+			if (Files.exists(Paths.get("chebi-cache")))
+			{
+				Files.readAllLines(Paths.get("chebi-cache")).parallelStream().forEach( line -> {
+					String[] parts = line.split("\t");
+					String oldChebiID = parts[0];
+					String newChebiID = parts[1];
+					String name = parts[2];
+					String formula = parts.length > 2 ? parts[3] : "";
+					chebiCache.put(oldChebiID, Arrays.asList(newChebiID, name, formula));
+				});
+			}
+			logger.debug("{} entries in the chebi-cache", chebiCache.size());
+		}
+		else
+		{
+			logger.info("useCache is FALSE - chebi-cache will NOT be read. ChEBI will be queried for ALL identifiers.");
+		}
+		return chebiCache;
+	}
+
+	/**
+	 * Extract a ChEBI entity from the cache, based on the identifier.
+	 * @param chebiCache - The cache from a file.
+	 * @param identifier - The chebi Identifier.
+	 * @return A ChEBI Entity that will be built from the data in the cache. It will only have: an Identifier, ChEBI Name, Formula.
+	 */
+	private Entity extractChEBIEntityFromCache(Map<String, List<String>> chebiCache, String identifier)
+	{
+		// Use the AccessibleEntity to set the formula.
+		Entity entity = new AccessibleEntity();
+		entity.setChebiId(chebiCache.get("CHEBI:"+identifier).get(0));
```

Comment:
I expect this wouldn't be a problem based on code flow, but would there ever be a worry of chebiCache not containing "CHEBI:`identifier`"?

Reply:
This method gets called when `chebiCache.containsKey("CHEBI:"+identifier)` evaluates to TRUE, so I'm not sure how this scenario could happen. If the logic surrounding the call to this method changed (such that there was no check before calling), then it _could_ be a problem. But currently, I don't think it could happen that the first time the cache is checked, it contains `CHEBI:${identifier}` and then it suddenly _doesn't_ on the second lookup.